libvips/foreign/cgifsave.c

```diff
@@ -303,6 +335,36 @@ vips_foreign_save_cgif_write_frame( VipsForeignSaveCgif *cgif )
 	frame_config.genFlags = 
 		CGIF_FRAME_GEN_USE_TRANSPARENCY | 
 		CGIF_FRAME_GEN_USE_DIFF_WINDOW;
+	frame_config.attrFlags = 0;
+
+	/* Switch per-frame alpha channel on.
+	 * Index 0 is used for pixels with alpha channel.
+	 */
+	if( cgif->has_transparency ) {
+		frame_config.attrFlags |= CGIF_FRAME_ATTR_HAS_ALPHA;
+		frame_config.transIndex = 0;
+	}
+
+	/* Do the transparency trick (only possible when no alpha channel present)
+	*/
+	if( cgif->frame_bytes_head ) {
+		if( !cgif->has_transparency ) {
+			const uint8_t transIndex = cgif->lp->count;
+			for(uint32_t i = 0; i < frame_rect->width * frame_rect->height; i++) {
+				if(cgif_cmppixel(frame_bytes + 4 * i, cgif->frame_bytes_head + 4 * i, cgif->maxerror) == 0) {
+					cgif->index[i] = transIndex;
+				} else {
+					memcpy(cgif->frame_bytes_head + 4 * i, frame_bytes + 4 * i, 4);
+				}
+			}
```

Comment:
The `*` inside the inner loop makes me uncomfortable, though the compiler ought to optimize this away. I think I'd still rework this as:

```C
                        const uint8_t trans_index = cgif->lp->count;
                        int n_pixels = frame_rect->width * frame_rect->height;
                        double maxerror = cgif->maxerror;

                        VipsPel *cur, *bef;
                        int i;

                        cur = frame_bytes;
                        bef = cgif->frame_bytes_head;
                        for( i = 0; i < n_pixels; i++ ) {
                                if( cgif_pixels_are_equal( cur, bef, maxerror ) )   
                                        cgif->index[i] = trans_index;
                                else {
                                        bef[0] = cur[0];
                                        bef[1] = cur[1];
                                        bef[2] = cur[2];
                                        bef[3] = cur[3];
                                }

                                cur += 4;
                                bef += 4;
                        }
```

Not all compilers will take out the very slow `memcpy()` for you.

You might as well make `maxerror` a double in the operation args as well.